src/shared_ptr.hpp

```diff
@@ -1,8 +1,109 @@
-# pragma once
+#pragma once
+
+#include <functional>
 
 template<typename T>
 class shared_ptr
 {
 public:
-  explicit shared_ptr(T*){}
+  shared_ptr()
+    : mValue(nullptr)
+    , mCtrlBlock(nullptr)
+  {
+  }
+
+  explicit shared_ptr(T* aPtr)
+    : mCtrlBlock(new control_block(1, 0))
+  {
+    mValue = aPtr;
+  }
+
+  explicit shared_ptr(T* aPtr, std::function<void(T*)> aDeleter)
+    : mCtrlBlock(new control_block(1, 0, aDeleter))
+  {
+    mValue = aPtr;
+  }
+
+  shared_ptr(const shared_ptr& aPtr) noexcept
+  {
+    if (aPtr.mCtrlBlock)
+    {
+      aPtr.mCtrlBlock->StrongRef++;
+    }
+
+    mValue = aPtr.mValue;
+    mCtrlBlock = aPtr.mCtrlBlock;
+  }
+
+  shared_ptr& operator=(const shared_ptr<T>& aPtr) noexcept
+  {
+    if (mCtrlBlock == aPtr.mCtrlBlock && mValue == aPtr.mValue)
+    {
+      return *this;
+    }
+
+    // assuming if mValue != nullptr means that mCtrlBlock can't be nullptr as well
+    if (mValue)
+    {
+      if (mCtrlBlock->StrongRef <= 1)
+      {
+        mCtrlBlock->Deleter ? mCtrlBlock->Deleter(mValue) : delete mValue;
+      }
+      else
+      {
+        mCtrlBlock->StrongRef--;
+      }
+    }
+
+    aPtr.mCtrlBlock->StrongRef++;
+    mValue = aPtr.mValue;
+    mCtrlBlock = aPtr.mCtrlBlock;
+    return *this;
+  }
+
+  T& operator*() const noexcept { return *mValue; }
+  T* operator->() const noexcept { return mValue; }
+  explicit operator bool() const noexcept { return mValue ? true : false; }
+
+  ~shared_ptr()
+  {
+    if (mValue)
+    {
+      if (mCtrlBlock->StrongRef <= 1)
+      {
+        mCtrlBlock->Deleter ? mCtrlBlock->Deleter(mValue) : delete mValue;
+        delete mCtrlBlock;
```

Comment:
deleting ctrl block after removing all strong references is wondering why we need week count at all